test/unit/BPool/BPool_Unbind.t.sol

```diff
@@ -0,0 +1,116 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.25;
+
+import {BPoolBase} from './BPoolBase.sol';
+
+import {IERC20} from '@openzeppelin/contracts/token/ERC20/IERC20.sol';
+import {IBPool} from 'interfaces/IBPool.sol';
+
+contract BPoolUnbind is BPoolBase {
+  address public secondToken = makeAddr('secondToken');
+
+  function setUp() public virtual override {
+    super.setUp();
+    vm.mockCall(secondToken, abi.encodePacked(IERC20.transferFrom.selector), abi.encode());
+  }
+
+  function test_RevertWhen_ReentrancyLockIsSet() external {
+    bPool.call__setLock(_MUTEX_TAKEN);
+    vm.expectRevert(IBPool.BPool_Reentrancy.selector);
+    // it should revert
+    bPool.unbind(token);
+  }
+
+  function test_RevertWhen_CallerIsNOTController(address _caller) external {
+    // it should revert
+    vm.assume(_caller != deployer);
+    vm.prank(_caller);
+    vm.expectRevert(IBPool.BPool_CallerIsNotController.selector);
+    bPool.unbind(token);
+  }
+
+  modifier whenCallerIsController() {
+    vm.startPrank(deployer);
+    _;
+  }
+
+  function test_RevertWhen_TokenIsNotBound() external whenCallerIsController {
+    vm.expectRevert(IBPool.BPool_TokenNotBound.selector);
+    // it should revert
+    bPool.unbind(token);
+  }
+
+  modifier whenTokenIsBound() {
+    _setRecord(token, IBPool.Record({bound: true, index: 0, denorm: tokenWeight}));
+    bPool.set__totalWeight(tokenWeight);
+    address[] memory tokens = new address[](1);
+    tokens[0] = token;
+    bPool.set__tokens(tokens);
+    _;
+  }
+
+  function test_RevertWhen_PoolIsFinalized() external whenCallerIsController whenTokenIsBound {
+    bPool.set__finalized(true);
+    // it should revert
+    vm.expectRevert(IBPool.BPool_PoolIsFinalized.selector);
+    bPool.unbind(token);
+  }
+
+  function test_WhenTokenIsLastOnTheTokensArray(uint256 _startTotalWeight)
+    external
+    whenCallerIsController
+    whenTokenIsBound
+  {
+    // NOTE: not fuzzing the existing tokens since modifier is always executed *before*
+    // what I could do for a setup here
+    _startTotalWeight = bound(_startTotalWeight, 0, MAX_TOTAL_WEIGHT - tokenWeight);
+    // add the weight of the already-bound token
+    bPool.set__totalWeight(bPool.getTotalDenormalizedWeight() + _startTotalWeight);
```

Comment:
hmmm, i understand you're fuzzing `_startTotalWeight` and that's why you need to do it here, but, perhaps it'd be better to just do funny-value-weight, set it on the modifier, and not have to declare it here.

also, not liking doing the `bPool.getTotalDenormalizedWeight()` call, we could know beforehand what's the value of it

Reply:
addressing in #140